libgnucash/app-utils/gnc-option.hpp

```diff
@@ -0,0 +1,254 @@
+#ifndef SWIG //SWIG pulls in GncOwner from swig-engine.
+struct _gncOwner;
+using GncOwner = _gncOwner;
+#endif
+struct _QofQuery;
+using QofQuery = _QofQuery;
+struct QofInstance_s;
+using QofInstance = QofInstance_s;
+template <typename ValueType> class GncOptionValue;
+class GncOptionAccountListValue;
+class GncOptionAccountSelValue;
+class GncOptionMultichoiceValue;
+template <typename ValueType> class GncOptionRangeValue;
+template <typename ValueType> class GncOptionValidatedValue;
+class GncOptionDateValue;
+
+template <typename T>
+struct is_OptionClassifier
+{
+    static constexpr bool value =
+        std::is_base_of_v<OptionClassifier, std::decay_t<T>>;
+};
+
+template <typename T> inline constexpr bool
+is_OptionClassifier_v = is_OptionClassifier<T>::value;
+
+template <typename T, typename U>
+struct is_same_decayed
+{
+    static constexpr bool value = std::is_same_v<std::decay_t<T>,
+                                                 std::decay_t<U>>;
+};
+
+template <typename T, typename U> inline constexpr bool
+is_same_decayed_v = is_same_decayed<T, U>::value;
+
+template <typename T>
+struct is_RangeValue
+{
+    static constexpr bool value =
+         (is_same_decayed_v<T, GncOptionRangeValue<int>> ||
+          is_same_decayed_v<T, GncOptionRangeValue<double>>);
+};
+
+template <typename T> inline constexpr bool
+is_RangeValue_v = is_RangeValue<T>::value;
+
+
+using GncOptionVariant = std::variant<GncOptionValue<std::string>,
+                                      GncOptionValue<bool>,
+                                      GncOptionValue<int64_t>,
+                                      GncOptionValue<const QofInstance*>,
+                                      GncOptionValue<const QofQuery*>,
+                                      GncOptionValue<const GncOwner*>,
+                                      GncOptionValue<SCM>,
+                                      GncOptionAccountListValue,
+                                      GncOptionAccountSelValue,
+                                      GncOptionMultichoiceValue,
+                                      GncOptionRangeValue<int>,
+                                      GncOptionRangeValue<double>,
+                                      GncOptionValidatedValue<const QofInstance*>,
+                                      GncOptionValidatedValue<const QofQuery*>,
+                                      GncOptionDateValue>;
+
+using GncOptionVariantPtr = std::unique_ptr<GncOptionVariant>;
+
+enum class GncOptionMultichoiceKeyType
+{
+    SYMBOL,
+    STRING,
+    NUMBER,
+};
+
+/** @class GncOption
+ *  @brief Represents the public interface for an option.
+ *  Polymorphism is provided by a std::variant member containing GncOptionValue
+ *  types.
+*/
+
+class GncOption
+{
+public:
+    template <typename OptionType,
+              typename std::enable_if_t<is_OptionClassifier_v<OptionType>,
+                               int>  = 0>
```

Comment:
I need some help  understanding this construct as well. From what I gather so far the std::enable_if_t bit will add an int typename to the template bits if OptionType is a valid type of options according to is_OptionClassifier_v. I'm lost at the addition of `= 0` there though.
The whole point is probably to add or remove this particular constructor from the GncOption derived class at compile time depending on whether that derived class matches certain type traits. The details of how this works exactly currently escapes me.

Reply:
`= 0` is the default value, provided so that specializations can be on only OptionType, e.g. `GncOption<GncOptionValue>(...)` instead of `GncOption<GncOptionValue, 42>(...)`.

Reply:
> The whole point is probably to add or remove this particular constructor from the GncOption derived class at compile time depending on whether that derived class matches certain type traits. The details of how this works exactly currently escapes me.

The details are encapsulated in the acronym SFINAE, Substitution Failure Is Not An Error. Just like GnuCash's scheduled transaction template transaction, a template isn't compilable code: It's code that the compiler can turn into compilable code if it encounters a call to a function with parameters that can be interpreted to match the template parameters. Thus
```
template <typename T> T add (T a, T b) { ... }
```
can generate code that will make
```
auto foo = add (31, 11);
auto bar = add (29.5, 12.5);
```
do the right thing.

The compiler-generated functions `int add (int, int)` and `float add (float, float)` are called specializations. That's the S in SFINAE.

 Where things get tricky is if `add` is implemented with `operator+`. If you don't tell the compiler otherwise it will cheerfully try to create `char* add (char*, char*)` which will not do what you want, so you would need a different template. You can make an explicit specialization,
```
template<>char* add<char*>(char* a, char* b)
{ 
    char* retval = malloc(strlen(a) + strlen(b) +1);
    memset (retval, 0, sizeof(retval));
    strcat (retval, a);
    strcat (retval, b);
    return retval;
}
```
But that has to be done for every single type that doesn't use operator+ the way you want. Using `std::enable_if` constrains the template so that only types that satisfy the conditions passed to it will cause the compiler to generate a specialization. If the type doesn't satisfy those conditions then the compiler can't make a specialization of the template. That's specialization failure and it's not an error; the compiler will look for another template to try. Of course it's an error if there's no template or function declaration that matches what it's looking for, in which case it will complain that it's an error to use a function without declaring it first and proceed to tell you all of the templates it tried to use and why it couldn't use each of them.

Reply:
Ok. Rougly understood now what std::enable_if_t does. I realize there's a lot more subtlety in template parameters than I thought. I'll learn as we go. Metaprogramming here we come... :smiley: